app/gui2/src/components/GraphNode.vue

```diff
@@ -262,18 +297,55 @@ function handleClick(e: PointerEvent) {
   }
 }
 
-const isCircularMenuVisible = ref(false)
-const isAutoEvaluationDisabled = ref(false)
-const isDocsVisible = ref(false)
-const isVisualizationVisible = ref(false)
+function updatePreprocessor(module: string, method: string, ...args: string[]) {
+  visualizationConfiguration.value = {
+    executionContextId: props.executionContextId,
+    visualizationModule: props.mainModule,
+    expression: { module, definedOnType: module, name: method },
+    ...(args.length !== 0 ? { positionalArgumentsExpressions: args } : {}),
+  }
+}
 
-const visualizationType = ref('Scatterplot')
-const visualization = shallowRef<Visualization>()
+watch(isVisualizationVisible, async (visible) => {
+  if (!visible) {
+    return
+  }
+  // TODO: should the old vis be detached?
+  await props.languageServer.attachVisualization(
+    visualizationId.value,
+    props.node.rootSpan.id,
+    visualizationConfiguration.value ?? {
+      executionContextId: props.executionContextId,
+      ...DEFAULT_VISUALIZATION_CONFIGURATION,
+    },
+  )
+})
 
-const queuedVisualizationData = computed<{}>(() =>
-  visualizationStore.sampleData(visualizationType.value),
-)
-const visualizationData = ref<{}>({})
+watch(visualizationConfiguration, async (config) => {
+  if (config != null && isVisualizationVisible.value) {
+    await props.languageServer.modifyVisualization(visualizationId.value, config)
+  }
+})
+
+function onVisualizationUpdate(vizUpdate: VisualizationUpdate) {
+  const json = vizUpdate.dataString()
+  const newData = json != null ? JSON.parse(json) : undefined
+  visualizationData.value = newData
+}
+
+onMounted(() => {
+  props.dataServer.on(
+    `${OutboundPayload.VISUALIZATION_UPDATE}:${visualizationId.value}`,
+    onVisualizationUpdate,
+  )
+})
+
+onUnmounted(() => {
+  props.dataServer.off(
+    `${OutboundPayload.VISUALIZATION_UPDATE}:${visualizationId.value}`,
+    onVisualizationUpdate,
+  )
+})
```

Comment:
As before, I don't like that this low level communication is done directly in the vue component's code. I think here we should at most have something like:
```ts
const visualizationData = useVisualizationData(visualizationConfigRef)
```